src/utility/connector.cpp

```diff
@@ -166,22 +166,32 @@ void connector::handle_resolve(const boost_code& ec, asio::iterator iterator,
         return;
     }
 
-    const auto timeout = settings_.connect_timeout();
-    const auto timer = std::make_shared<deadline>(pool_, timeout);
-    const auto socket = std::make_shared<network::socket>(pool_);
-
-    // Retain a socket reference until connected, allowing connect cancelation.
-    pending_.store(socket);
-
-    // Manage the socket-timer race, terminating if either fails.
-    const auto handle_connect = synchronize(handler, 1, NAME, mode);
-
-    // This is branch #1 of the connnect sequence.
-    timer->start(
-        std::bind(&connector::handle_timer,
-            shared_from_this(), _1, socket, handle_connect));
-
-    safe_connect(iterator, socket, timer, handle_connect);
+    auto do_connecting = [this, &handler](asio::iterator resolver_iterator)
+    {
+        const auto timeout = settings_.connect_timeout();
+        const auto timer = std::make_shared<deadline>(pool_, timeout);
+        const auto socket = std::make_shared<network::socket>(pool_);
+    
+        // Retain a socket reference until connected, 
+        // allowing connect cancelation.
+        pending_.store(socket);
+    
+        // Manage the socket-timer race, terminating if either fails.
+        const auto handle_connect = synchronize(handler, 1, NAME, mode);
+    
+        // This is branch #1 of the connnect sequence.
+        timer->start(
+            std::bind(&connector::handle_timer,
+                shared_from_this(), _1, socket, handle_connect));
+    
+        safe_connect(resolver_iterator, socket, timer, handle_connect);
+    };
+    
+    // Get all hosts under one DNS record.
+    for (asio::iterator end; iterator != end; ++iterator)
+    {  
+        do_connecting(iterator);
```

Comment:
This will continue to loop following the first connection, yet the completion handler will have already been called. This extends the thread execution unnecessarily after connect. I reccomend converting safe connect and this closure to bool and using success as a second loop termination condition.